[IN-83] Login to app as an admin (main app and backoffice)

adamjorge opened the pull request:
Linked to this ticket: https://instamint-deva.atlassian.net/browse/IN-83

adamjorge committed:
chore: starting with next auth config

adamjorge committed:
feat: working middleware

adamjorge committed:
feat(middleware): protect admin routes

adamjorge committed:
feat: working auth with custom credentials login page

adamjorge committed:
fix: bcrypt dependency

adamjorge committed:
ci: create .env file

adamjorge committed:
ci: update node version

ci: debug .env

ci: debug .env

ci: debug .env

ci: debug .env

adamjorge committed:
ci: debug vercel preview build

adamjorge committed:
ci: debug vercel preview build

adamjorge committed:
ci(package.json): node engine

ci(package.json): node engine

adamjorge committed:
feat: connection form + bcrypt decode

adamjorge committed:
chore: refacto on sign in form + changed prisma conf

adamjorge committed:
chore: sign out from server

adamjorge committed:
chore(prisma model): update to handle roles in session

adamjorge committed:
feat(login): admin protected route and user login logic

adamjorge committed:
chore: same layout for auth root logic

adamjorge committed:
chore: update seeds

jimmy-martin commented on .env.example:
```diff
@@ -1 +1,24 @@
\ No newline at end of file
+# Created by Vercel CLI
+AUTH_SECRET=""
+DATABASE_URL=""
```
You can pre-fill `DATABASE_URL` with the basic connection string as it will be easier to modify it afterwards

jimmy-martin commented on prisma/schema.prisma:
```diff
@@ -26,12 +65,27 @@ model Hashtag {
+  accounts      Account[]
+  sessions      Session[]
+  minterId      Int
+  Minter        Minter    @relation(fields: [minterId], references: [id])
```
use lowercase instead : 
`minter Minter @relation(fields: [minterId], references: [id])`

jimmy-martin commented on prisma/schema.prisma:
```diff
@@ -40,14 +94,10 @@ model Minter {
   Comment             Comment[]
   OriginalContent     OriginalContent[]
   Report              Report[]
+  User                User[]
```
Same here, you can use lowercase and dont forget to use plural (update all the prisma schema if you can as we saw together): 

```
comments             Comment[]
originalContents     OriginalContent[]
reports              Report[]
users                User[]
```

jimmy-martin commented on src/app/[locale]/layout.tsx:
```diff
@@ -68,7 +70,19 @@ export default function PortalLayout({
+                }}
+              >
+                <Button type="submit" className="bg-red-500">
+                  Sign out
```
you can add a translation for "Sign out"

jimmy-martin commented on src/lib/auth.ts:
Is it better to use eslint disabling rules on the entire file instead of multiplying `// eslint-disable-next-line new-cap` ?

adamjorge replied:
I can change that, it was just to show that i'm respecting linter practices for the rest of the file

jimmy-martin commented on src/lib/db.ts:
```diff
@@ -13,10 +13,10 @@ declare global {
 }
 
-const prisma = global.prismaGlobal || prismaSingleton()
+const prisma = global.prismaGlobal ?? prismaSingleton()
```
Why this change ?

adamjorge replied:
Same use case, I just took it from the official docs because I tried other confs:

https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-dev-practices

jimmy-martin commented on src/lib/db.ts:
```diff
@@ -13,10 +13,10 @@ declare global {
 
 export default prisma
 
-if (process.env.NODE_ENV !== "production" && !global.prismaGlobal) {
```
Why this change ? It was here to avoid to set global.prismaGlobal = prisma when it's already set.

adamjorge replied:
Same use case, I just took it from the official docs because I tried other confs:

https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-dev-practices

jimmy-martin commented on src/lib/query/minters/search.ts:
```diff
@@ -11,7 +11,6 @@ export async function searchMinters(search: string) {
       location: true
     },
     where: {
-      isAdmin: false,
```
do you want to list admins on search ?

adamjorge replied:
An admin is also considered as a Minter in our app but we can talk about that.
I you really want to not list admin, it means that the search query needs a join query to access isAdmin in User (changed Minter.isAdmin => User.isAdmin)

adamjorge committed:
refactor(translation + app structure): changed middleware and main routes

adamjorge committed:
fix(i18n): sign out translation

adamjorge committed:
test: updated TU and e2e tests for connection of admin user

adamjorge committed:
ci: updated secrets to access dev admin credentials

adamjorge committed:
ci: trying another conf for playwright

ci: trying another conf

ci: trying another conf

ci: trying another conf

ci: trying another conf

ci: trying another conf

ci: trying another conf

chore(playwright config): removed safari dependency because it's unstable for the moment

avetisk commented on prisma/schema.prisma:
```diff
@@ -1,12 +1,51 @@
+}
+
+model Session {
+  id           String   @id @default(cuid())
+  userId       String
+  expires      DateTime
+  sessionToken String   @unique
+  accessToken  String   @unique
+  createdAt    DateTime @default(now())
+  updatedAt    DateTime @updatedAt
+  user         User     @relation(fields: [userId], references: [id])
```
Not a good idea to handle sessions in a SQL database.

It's not scalable.

avetisk commented on prisma/schema.prisma:
```diff
@@ -1,12 +1,51 @@
+  user         User     @relation(fields: [userId], references: [id])
+}
+
+model VerificationRequest {
+  id         String   @id @default(cuid())
+  identifier String
+  token      String   @unique
+  expires    DateTime
+  createdAt  DateTime @default(now())
+  updatedAt  DateTime @updatedAt
+
+  @@unique([identifier, token])
+}
+
```
Just use JWT or even better PASETO token for this.

You can always keep in a KV in memory stores those tokens if you want to double validate them to ensure single usage.

avetisk commented on src/app/[locale]/(portal)/layout.tsx:
```diff
@@ -0,0 +1,31 @@
+
+  return (
+    <div>
+      <form
+        action={async () => {
+          "use server"
+          await signOut()
+        }}
+      >
```
Server Action === PHP === 🙅‍♂️

avetisk commented on src/app/[locale]/(portal)/page.tsx:
```diff
@@ -0,0 +1,5 @@
+"use client"
+
+export default function Home() {
+  return <></>
```
Just return `null` 🤣

adamjorge replied:
Updated with the feed of the minter in this PR: https://github.com/adammehdaoui/instamint/pull/25

avetisk commented on src/lib/auth.ts:
```diff
@@ -0,0 +1,66 @@
+    })
+  ],
+  pages: {
+    signIn: "/en/login"
```
What about i18n?

adamjorge replied:
Changes made in this PR: https://github.com/adammehdaoui/instamint/pull/18
("/en/login" => "/login")
